src/features/side_by_side_wrap.rs

```diff
@@ -0,0 +1,1115 @@
+use crate::features::line_numbers;
+use crate::features::line_numbers::SideBySideLineWidth;
+use crate::features::side_by_side::available_line_width;
+use crate::features::side_by_side::line_is_too_long;
+use crate::features::side_by_side::PanelSide::*;
+use crate::minusplus::*;
+use crate::style::Style;
+
+/// See [`wrap_line`] for documentation.
+#[derive(Clone, Debug)]
+pub struct WrapConfig {
+    pub left_symbol: String,
+    pub right_symbol: String,
+    pub right_prefix_symbol: String,
+    // In fractions of 1000 so that a >100 wide panel can
+    // still be configured down to a single character.
+    pub use_wrap_right_permille: usize,
+    pub max_lines: usize,
+    pub inline_hint_syntect_style: SyntectStyle,
+}
+
+/// Wrap the given `line` if it is longer than `line_width`. Wrap to at most
+/// [Config::WrapConfig::max_lines](WrapConfig::max_lines) lines,
+/// then truncate again - but never truncate if it is `0`. Place
+/// [left_symbol](WrapConfig::left_symbol) at the end of wrapped lines.
+/// If wrapping results in only *one* extra line and if the width of the wrapped
+/// line is less than [use_wrap_right_permille](WrapConfig::use_wrap_right_permille)
+/// then right-align the second line and use the symbols
+/// [right_symbol](WrapConfig::right_symbol) and
+/// on the next line [right_prefix_symbol](WrapConfig::right_prefix_symbol).
+/// The inserted characters will follow the
+/// [inline_hint_syntect_style](WrapConfig::inline_hint_syntect_style).
+///
+/// The input `line` is expected to start with an (ultimately not printed) `+`, `-` or ` ` prefix.
+/// The prefix `_` is also added to the start of wrapped lines.
+pub fn wrap_line<'a, I, S>(
+    config: &'a Config,
+    line: I,
+    line_width: usize,
+    fill_style: &S,
+    inline_hint_style: &Option<S>,
+) -> Vec<Vec<(S, &'a str)>>
+where
+    I: IntoIterator<Item = (S, &'a str)> + std::fmt::Debug,
+    <I as IntoIterator>::IntoIter: DoubleEndedIterator,
+    S: Copy + Default + std::fmt::Debug,
+{
+    let mut result = Vec::new();
+
+    let wrap_config = &config.wrap_config;
+
+    // Symbol which:
+    //  - represents the additional "+/-/ " prefix on the unwrapped input line, its
+    //    length is added to the line_width.
+    //  - can be more prominent than a space because syntax highlighting has already
+    //    been done.
+    //  - is added at the beginning of wrapped lines so the wrapped lines also have
+    //    a prefix (which is not printed).
+    const LINEPREFIX: &str = "_";
+    assert_eq!(LINEPREFIX.len(), INLINE_SYMBOL_WIDTH_1); // (args are const, optimized out)
+
+    let max_len = line_width + LINEPREFIX.len();
+
+    // The current line being assembled from the input to fit exactly into the given width.
+    // A somewhat leaky abstraction as the fields are also accessed directly.
+    struct CurrLine<'a, S: Default> {
+        line_segments: Vec<(S, &'a str)>,
+        len: usize,
+    }
+    impl<'a, S: Default> CurrLine<'a, S> {
```

Comment:
For my own education -- is the reason this type and its associated methods are declared inside the function definition that we want to ensure that the type parameters supplied to the generic types are "tied together", i.e. that `CurrLine<S>` is only defined for `S` such that  `wrap_line<S>` is defined? Is declaring `CurrLine` at the top level (along with `LINEPREFIX`) attractive to you at all? (the advantage I see would be to keep the function body shorter and simpler)

(By the way, I don't use Rust professionally, and I've never really used C++ or another language with generics, so you are going to be much better informed about the facts and what's standard style than I am.)